File: frontend/src/lib/agentcore-invoke.ts
import { readFile } from 'fs/promises';
import path from 'path';
import { randomUUID } from 'crypto';
import {
  BedrockAgentCoreClient,
  InvokeAgentRuntimeCommand,
  StopRuntimeSessionCommand,
} from '@aws-sdk/client-bedrock-agentcore';
import { NodeHttpHandler } from '@smithy/node-http-handler';
import { loadBackendEnv } from './backend-env';
import { getBackendRoot } from './repo-root';

export interface AgentCoreInvokeResult {
  status: 'success' | 'error';
  text?: string;
  error?: string;
  agentName: string;
  runtimeArn?: string;
  runtimeSessionId?: string;
  response?: unknown;
}

/** AgentCore requires session ids 33–256 chars; UUID hex + trailing digit satisfies that. */
export function newRuntimeSessionId(): string {
  return `${randomUUID().replace(/-/g, '')}0`;
}

function a2aMessageSendPayload(messageText: string): Uint8Array {
  const body = {
    jsonrpc: '2.0',
    id: randomUUID().replace(/-/g, ''),
    method: 'message/send',
    params: {
      message: {
        role: 'user',
        messageId: randomUUID().replace(/-/g, ''),
        parts: [{ kind: 'text', text: messageText }],
      },
    },
  };
  return new TextEncoder().encode(JSON.stringify(body));
}

export function extractTextFromA2aJsonrpc(data: unknown): string {
  if (!data || typeof data !== 'object') return JSON.stringify(data, null, 2);
  const record = data as Record<string, unknown>;

  if (record.error) {
    const err = record.error;
    if (err && typeof err === 'object') {
      const e = err as Record<string, unknown>;
      return `A2A error ${e.code}: ${e.message}`;
    }
    return `A2A error: ${String(err)}`;
  }

  const result = record.result;
  if (!result || typeof result !== 'object') return JSON.stringify(data, null, 2);

  const texts: string[] = [];
  const res = result as Record<string, unknown>;

  for (const artifact of (res.artifacts as unknown[]) ?? []) {
    if (!artifact || typeof artifact !== 'object') continue;
    for (const part of ((artifact as Record<string, unknown>).parts as unknown[]) ?? []) {
      if (!part || typeof part !== 'object') continue;
      const p = part as Record<string, unknown>;
      if (p.kind === 'text' && p.text) texts.push(String(p.text));
    }
  }

  const message = res.message;
  if (message && typeof message === 'object') {
    for (const part of ((message as Record<string, unknown>).parts as unknown[]) ?? []) {
      if (!part || typeof part !== 'object') continue;
      const p = part as Record<string, unknown>;
      if (p.kind === 'text' && p.text) texts.push(String(p.text));
    }
  }

  for (const entry of (res.history as unknown[]) ?? []) {
    if (!entry || typeof entry !== 'object') continue;
    const msg = (entry as Record<string, unknown>).message ?? entry;
    if (!msg || typeof msg !== 'object') continue;
    for (const part of ((msg as Record<string, unknown>).parts as unknown[]) ?? []) {
      if (!part || typeof part !== 'object') continue;
      const p = part as Record<string, unknown>;
      if (p.kind === 'text' && p.text) texts.push(String(p.text));
    }
    const role = (msg as Record<string, unknown>).role;
    const content = (msg as Record<string, unknown>).content;
    if (role === 'assistant' && Array.isArray(content)) {
      for (const block of content) {
        if (block && typeof block === 'object' && (block as Record<string, unknown>).text) {
          texts.push(String((block as Record<string, unknown>).text));
        }
      }
    }
  }

  if (texts.length) return texts.join('\n').trim();

  const status = res.status;
  if (typeof status === 'string' && status) return status;

  return '';
}

let _client: BedrockAgentCoreClient | null = null;
let _clientTimeoutMs = 0;

function agentCoreClient(readTimeoutMs: number): BedrockAgentCoreClient {
  loadBackendEnv();
  if (_client && _clientTimeoutMs >= readTimeoutMs) return _client;

  const region = process.env.AWS_REGION?.trim() || 'us-east-2';
  const profile = process.env.AWS_PROFILE?.trim();
  if (profile) process.env.AWS_PROFILE = profile;

  _clientTimeoutMs = readTimeoutMs;
  _client = new BedrockAgentCoreClient({
    region,
    maxAttempts: 2,
    requestHandler: new NodeHttpHandler({
      connectionTimeout: 60_000,
      requestTimeout: readTimeoutMs,
    }),
  });
  return _client;
}

export async function loadRuntimeArn(agentName: string): Promise<string | null> {
  const meta = await loadAgentRuntimeMeta(agentName);
  return meta.runtimeArn;
}

export async function loadAgentRuntimeMeta(
  agentName: string,
): Promise<{ deployed: boolean; runtimeArn: string | null }> {
  const file = path.join(getBackendRoot(), 'config', 'agentcore', 'runtimes.json');
  try {
    const data = JSON.parse(await readFile(file, 'utf-8')) as {
      agents?: Record<string, { deployed?: boolean; runtimeArn?: string }>;
    };
    const entry = data.agents?.[agentName];
    const runtimeArn = entry?.runtimeArn?.trim() || null;
    return { deployed: entry?.deployed === true && Boolean(runtimeArn), runtimeArn };
  } catch {
    return { deployed: false, runtimeArn: null };
  }
}

/** Invoke a deployed AgentCore runtime via boto3-compatible SDK (A2A message/send). */
export async function invokeAgentRuntimeA2a(
  agentName: string,
  messageText: string,
  options?: { timeoutSec?: number; runtimeSessionId?: string },
): Promise<AgentCoreInvokeResult> {
  const timeoutSec = options?.timeoutSec ?? 900;
  const readTimeoutMs = Math.max(timeoutSec * 1000, 60_000);

  const arn = await loadRuntimeArn(agentName);
  if (!arn) {
    return {
      status: 'error',
      agentName,
      error: `No runtimeArn for ${agentName} in config/agentcore/runtimes.json`,
    };
  }

  const sessionId = options?.runtimeSessionId?.trim() || newRuntimeSessionId();
  const client = agentCoreClient(readTimeoutMs);

  try {
    const response = await client.send(
      new InvokeAgentRuntimeCommand({
        agentRuntimeArn: arn,
        runtimeSessionId: sessionId,
        payload: a2aMessageSendPayload(messageText),
        contentType: 'application/json',
        accept: 'application/json',
      }),
    );

    const statusCode = response.statusCode ?? 200;
    let bodyText = '';
    if (response.response) {
      const raw = response.response;
      if (typeof raw === 'string') {
        bodyText = raw;
      } else if (raw instanceof Uint8Array) {
        bodyText = new TextDecoder().decode(raw);
      } else if (typeof (raw as { transformToString?: () => Promise<string> }).transformToString === 'function') {
        bodyText = await (raw as { transformToString: () => Promise<string> }).transformToString();
      }
    }

    if (statusCode >= 400) {
      return {
        status: 'error',
        agentName,
        runtimeArn: arn,
        runtimeSessionId: sessionId,
        error: `HTTP ${statusCode}: ${bodyText.slice(0, 500)}`,
      };
    }

    let parsed: unknown = {};
    if (bodyText.trim()) {
      try {
        parsed = JSON.parse(bodyText);
      } catch {
        return {
          status: 'success',
          agentName,
          runtimeArn: arn,
          runtimeSessionId: sessionId,
          text: bodyText,
          response: { raw: bodyText },
        };
      }
    }

    return {
      status: 'success',
      agentName,
      runtimeArn: arn,
      runtimeSessionId: sessionId,
      text: extractTextFromA2aJsonrpc(parsed),
      response: parsed,
    };
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return {
      status: 'error',
      agentName,
      runtimeArn: arn,
      runtimeSessionId: sessionId,
      error: message,
    };
  }
}

/** Hard-stop an AgentCore runtime session (cuts further pipeline work on that session). */
export async function stopRuntimeSession(
  agentName: string,
  runtimeSessionId: string,
  runtimeArn?: string | null,
): Promise<{ ok: boolean; error?: string }> {
  const sessionId = runtimeSessionId.trim();
  if (!sessionId) return { ok: false, error: 'runtimeSessionId is required' };

  const arn = runtimeArn?.trim() || (await loadRuntimeArn(agentName));
  if (!arn) {
    return { ok: false, error: `No runtimeArn for ${agentName}` };
  }

  try {
    const client = agentCoreClient(60_000);
    await client.send(
      new StopRuntimeSessionCommand({
        agentRuntimeArn: arn,
        runtimeSessionId: sessionId,
        qualifier: 'DEFAULT',
      }),
    );
    return { ok: true };
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    // Already gone / unknown session is still a successful cancel from the UI POV.
    if (/not.?found|does not exist|ResourceNotFound|ConflictException/i.test(message)) {
      return { ok: true, error: message };
    }
    return { ok: false, error: message };
  }
}

File: backend/agents/_shared/agentcore_invoke.py
"""Invoke Bedrock AgentCore Runtimes by ARN (via boto3)."""

from __future__ import annotations

import json
import logging
import os
from pathlib import Path
from typing import Any
from uuid import uuid4

from .artifact_store import repo_root

logger = logging.getLogger(__name__)

RUNTIMES_CONFIG = repo_root() / "config" / "agentcore" / "runtimes.json"


def load_runtime_arn(agent_name: str) -> str | None:
    """Resolve specialist runtime ARN.

    Priority:
    1. ``AGENTCORE_PEER_RUNTIME_ARNS`` JSON map (``{"product-agent":"arn:..."}``)
       — used by demo orchestrator so peers stay env-pinned without rebuilding.
    2. ``AGENTCORE_RUNTIMES_CONFIG`` file (e.g. ``config/agentcore/runtimes.demo.json``)
    3. Default ``config/agentcore/runtimes.json``
    """
    raw_peers = os.getenv("AGENTCORE_PEER_RUNTIME_ARNS", "").strip()
    if raw_peers:
        try:
            peers = json.loads(raw_peers)
        except json.JSONDecodeError:
            logger.warning("AGENTCORE_PEER_RUNTIME_ARNS is not valid JSON")
            peers = None
        if isinstance(peers, dict):
            arn = str(peers.get(agent_name) or "").strip()
            if arn:
                return arn

    override = os.getenv("AGENTCORE_RUNTIMES_CONFIG", "").strip()
    if override:
        cfg_path = Path(override)
        if not cfg_path.is_absolute():
            cfg_path = repo_root() / cfg_path
    else:
        cfg_path = RUNTIMES_CONFIG

    if not cfg_path.is_file():
        return None
    try:
        data = json.loads(cfg_path.read_text(encoding="utf-8"))
    except json.JSONDecodeError:
        return None
    entry = (data.get("agents") or {}).get(agent_name) or {}
    arn = str(entry.get("runtimeArn") or "").strip()
    return arn or None


def _a2a_message_send_payload(message_text: str) -> bytes:
    body = {
        "jsonrpc": "2.0",
        "id": uuid4().hex,
        "method": "message/send",
        "params": {
            "message": {
                "role": "user",
                "messageId": uuid4().hex,
                "parts": [{"kind": "text", "text": message_text}],
            }
        },
    }
    return json.dumps(body).encode("utf-8")


def extract_text_from_a2a_jsonrpc(data: dict[str, Any]) -> str:
    """Pull agent text from an A2A JSON-RPC response body."""
    if data.get("error"):
        err = data["error"]
        if isinstance(err, dict):
            return f"A2A error {err.get('code')}: {err.get('message')}"
        return f"A2A error: {err}"

    result = data.get("result")
    if not isinstance(result, dict):
        return json.dumps(data, indent=2)

    texts: list[str] = []
    for artifact in result.get("artifacts") or []:
        if not isinstance(artifact, dict):
            continue
        for part in artifact.get("parts") or []:
            if isinstance(part, dict) and part.get("kind") == "text":
                text = part.get("text")
                if text:
                    texts.append(str(text))

    message = result.get("message")
    if isinstance(message, dict):
        for part in message.get("parts") or []:
            if isinstance(part, dict) and part.get("kind") == "text":
                text = part.get("text")
                if text:
                    texts.append(str(text))

    if texts:
        return "\n".join(texts)
    return json.dumps(result, indent=2)


def _read_response_body(raw: Any) -> str:
    if raw is None:
        return ""
    if hasattr(raw, "read"):
        raw = raw.read()
    if isinstance(raw, bytes):
        return raw.decode("utf-8", errors="replace")
    return str(raw)


def invoke_agent_runtime_a2a(
    agent_name: str,
    message_text: str,
    *,
    timeout: int = 600,  # noqa: ARG001 — boto3 uses botocore read timeout from config
) -> dict[str, Any]:
    """Call invoke_agent_runtime with an A2A message/send JSON-RPC body."""
    arn = load_runtime_arn(agent_name)
    if not arn:
        return {
            "status": "error",
            "error": f"No runtimeArn for {agent_name} in {RUNTIMES_CONFIG}",
            "agent_name": agent_name,
        }

    import boto3
    from botocore.config import Config

    region = os.getenv("AWS_REGION", "us-east-2")
    client = boto3.client(
        "bedrock-agentcore",
        region_name=region,
        config=Config(read_timeout=timeout, connect_timeout=60, retries={"max_attempts": 1}),
    )
    session_id = uuid4().hex + "0"  # AgentCore requires runtimeSessionId length >= 33
    try:
        response = client.invoke_agent_runtime(
            agentRuntimeArn=arn,
            runtimeSessionId=session_id,
            payload=_a2a_message_send_payload(message_text),
            contentType="application/json",
            accept="application/json",
        )
    except Exception as exc:
        logger.exception("invoke_agent_runtime failed for %s", agent_name)
        return {
            "status": "error",
            "error": str(exc),
            "agent_name": agent_name,
            "runtime_arn": arn,
        }

    status_code = int(response.get("statusCode") or 200)
    body_text = _read_response_body(response.get("response"))
    if status_code >= 400:
        return {
            "status": "error",
            "error": f"HTTP {status_code}: {body_text[:500]}",
            "agent_name": agent_name,
            "runtime_arn": arn,
        }

    try:
        parsed = json.loads(body_text) if body_text.strip() else {}
    except json.JSONDecodeError:
        return {
            "status": "success",
            "response": {"raw": body_text},
            "text": body_text,
            "agent_name": agent_name,
            "runtime_arn": arn,
            "runtime_session_id": session_id,
        }

    text = extract_text_from_a2a_jsonrpc(parsed)
    return {
        "status": "success",
        "response": parsed,
        "text": text,
        "agent_name": agent_name,
        "runtime_arn": arn,
        "runtime_session_id": session_id,
    }


def should_use_agentcore_arn_invoke() -> bool:
    """Use boto3 invoke when running on AgentCore or when S3 pipeline mode is enabled."""
    flag = os.getenv("AGENTCORE_USE_ARN_INVOKE", "").strip().lower()
    if flag in ("1", "true", "yes", "on"):
        return True
    if flag in ("0", "false", "no", "off"):
        return False
    if os.getenv("AGENTCORE_AGENT", "").strip():
        return True
    from .artifact_store import is_s3_store

    return is_s3_store()

